src/util.ts

```diff
@@ -7,8 +7,165 @@ import { access } from 'fs/promises';
+     *
+     * The purpose of this property is to signal that new connections to the particular endpoint should not be
+     * attempted, for example because it uses an untrusted TLS certificate. Regular (unexpected) disconnects do not set
+     * the error property. This property is only reused per client connection.
+     */
+    error: string | null;
+    /**
+     * `true` if TLS should be established, `false` otherwise. Note that this property only describes if a connection
+     * should eventually be protected using TLS. To check if TLS has already been established, use
+     * {@link MitmproxyConnection.tlsEstablished}.
+     */
+    tls: boolean;
+    /** The TLS certificate list as sent by the peer. The first certificate is the end-entity certificate. */
+    certificateList: MitmproxyCertificate[];
+    /** The active cipher name as returned by OpenSSL's `SSL_CIPHER_get_name`. */
+    cipherName: string;
+    /** Ciphers accepted by the proxy server on this connection. */
+    ciphersClientToProxy: string[];
+    /** The active TLS version. */
+    tlsVersion: string | null;
+    /** The Server Name Indication (SNI) sent in the ClientHello. */
+    sni: string | null;
+    /** Timestamp of when the TCP SYN was received (client) or sent (server). */
+    timestampStart: number | null;
+    /** Timestamp of when the connection has been closed. */
+    timestampEnd: number | null;
+    /** Timestamp of when the TLS handshake has been completed successfully. */
+    timestampTlsSetup: number | null;
+    /** `true` if {@link MitmproxyConnection.state} is `OPEN`, `false` otherwise. */
+    connected: boolean;
+    /** `true` if TLS has been established, `false` otherwise. */
+    tlsEstablished: boolean;
+};
+
+/**
+ * Mitmproxy's event data for the `tls_start_client`, `tls_start_server`, and `tls_handshake` event hooks.
+ *
+ * @see https://docs.mitmproxy.org/stable/api/mitmproxy/tls.html#TlsData
+ */
+export type MitmproxyTlsData = {
+    /** Convenience alias for the client address in human-readable format (`<address>:<port>`). */
+    clientAddress: string;
+    /** Convenience alias for the server address in human-readable format (SNI hostname or `<address>:<port>`). */
+    serverAddress: string;
+
+    /** The client connection. */
+    client: MitmproxyClient;
+    /** The server connection. */
+    server: MitmproxyServer;
+    /** If set to `true`, indicates that it is a DTLS event. */
+    isDtls: boolean;
+};
+
+/**
+ * A mitmproxy client object, represents a connection between a client and mitmproxy.
+ *
+ * @see https://docs.mitmproxy.org/stable/api/mitmproxy/connection.html#Client
+ */
+export type MitmproxyClient = MitmproxyConnection & {
+    /** The certificate used by mitmproxy to establish TLS with the client. */
+    mitmCertificate: MitmproxyCertificate | null;
+    /** The proxy server type this client has been connecting to. */
+    proxyMode: string;
```

Comment:
No, it can't be. That is `client.proxy_mode.full_spec`. According to the docstring:

> The full proxy mode spec as entered by the user.

And while I guess we could try to construct a more strict type here, it would be a lot more complicated than `'regular' | 'wireguard'`:

```
  --mode MODE, -m MODE  The proxy server type(s) to spawn. Can be passed multiple times. Mitmproxy supports "regular" (HTTP), "transparent",
                        "socks5", "reverse:SPEC", and "upstream:SPEC" proxy servers. For reverse and upstream proxy modes, SPEC is host
                        specification in the form of "http[s]://host[:port]". You may append `@listen_port` or `@listen_host:listen_port` to
                        override `listen_host` or `listen_port` for a specific proxy mode. Features such as client playback will use the first
                        mode to determine which upstream server to use. May be passed multiple times.
```

Reply:
But I will update our docstring to make this clearer.